### ISSUE TITLE
Sub-meshes that share material are merged?

### ISSUE THREAD

**Issue by Stinkfist0** [IfcOpenShell/IfcOpenShell#104]
Toying around with layer set slicing. I have e.g. a wall of which material specification looks like this in Solibri Model Viewer:
![image](https://cloud.githubusercontent.com/assets/227876/16911628/412f6e2c-4ce8-11e6-9910-86de1d9f8689.png)

However, in the resulted DAE, the mesh has only two sub-meshes and the both outer wall parts (Gypsum Wall Board) are part of the same sub-mesh and share material usage, and I cannot choose different materials for inner and outer wall parts. Is this a bug in the layer set slicing code?

**Comment by aothms** [comment]
I would expect the layers to end up in different meshes. Do you have a small test file you can share?

**Comment by Stinkfist0** [comment]
I'll try to see if I can segregate a small test file. I also try to research a bit more, maybe check if there's difference between OBJ and DAE, etc.

**Comment by Stinkfist0** [comment]
Actually this issue can be reproduced with Duplex_A_20110907.ifc `: IfcConvert Duplex_A_20110907.ifc Duplex_A_20110907-layerset-sliced.dae --enable-layerset-slicing

In Solibri this wall has 7 materials:
![image](https://cloud.githubusercontent.com/assets/227876/16913096/f3a83cde-4cf0-11e6-892f-faab7b11ac8d.png)


But in Unity only 4
![image](https://cloud.githubusercontent.com/assets/227876/16913085/ec2e41c4-4cf0-11e6-952a-d2cb238a5079.png)

**Comment by aothms** [comment]
This appears to be an optimization happening inside unity. While the object indeed has 4 materials assigned to it, the mesh has 7 streams of triangles.

~~~
>>> ifcopenshell.guid.split(ifcopenshell.guid.expand("2O2Fr$t4X7Zf8NOew3FKau"))
'{9808fd7f-dc48-478e-9217-628e833d4938}'

<node id="product-9808fd7f-dc48-478e-9217-628e833d4938-body" name="product-9808fd7f-dc48-478e-9217-628e833d4938-body" type="NODE">
  <matrix>0 1 0 4.400000000000001 -1 0 0 -0.4170000000000139 0 0 1 3.100000000000378 0 0 0 1</matrix>
  <instance_geometry url="#representation-5397">
    <bind_material>
      <technique_common>
        <instance_material symbol="surface-style-3944-plasterboard" target="#surface-style-3944-plasterboard"/>
        <instance_material symbol="surface-style-3935-metal---stud-layer" target="#surface-style-3935-metal---stud-layer"/>
        <instance_material symbol="surface-style-3920-masonry---concrete-block" target="#surface-style-3920-masonry---concrete-block"/>
        <instance_material symbol="surface-style-3896-misc.-air-layers---air-space" target="#surface-style-3896-misc.-air-layers---air-space"/>
      </technique_common>
    </bind_material>
  </instance_geometry>
</node>

<geometry id="representation-5397">
<mesh>
  <source id="representation-5397-positions">
    <float_array id="representation-5397-positions-array" count="948">0 0.2750000000000004 2.900000000000012 6.248999999999988 0.2750000000000007 2.900000000001 0
    <technique_common>
      <accessor source="#representation-5397-positions-array" count="316" stride="3">
        <param name="X" type="float"/>
        <param name="Y" type="float"/>
        <param name="Z" type="float"/>
      </accessor>
    </technique_common>
  </source>
  <source id="representation-5397-normals">
    <float_array id="representation-5397-normals-array" count="948">0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 0 1 0 0
    <technique_common>
      <accessor source="#representation-5397-normals-array" count="316" stride="3">
        <param name="X" type="float"/>
        <param name="Y" type="float"/>
        <param name="Z" type="float"/>
      </accessor>
    </technique_common>
  </source>
  <vertices id="representation-5397-vertices">
    <input semantic="POSITION" source="#representation-5397-positions"/>
  </vertices>
  <triangles material="surface-style-3944-plasterboard" count="44">
    <input semantic="VERTEX" source="#representation-5397-vertices" offset="0"/>
    <input semantic="NORMAL" source="#representation-5397-normals" offset="1"/>
    <p>12 12 0 0 1 1 12 12 1 1 2 2 12 12 15 15 0 0 14 14 15 15 12 12 13 13 14 14 12 12 3 3 12 12 2 2 11 11 3 3 4 4 11 11 12 12 3 3 9 9 10 10 11 11 9 9
  </triangles>
  <triangles material="surface-style-3935-metal---stud-layer" count="28">
    <input semantic="VERTEX" source="#representation-5397-vertices" offset="0"/>
    <input semantic="NORMAL" source="#representation-5397-normals" offset="1"/>
    <p>65 65 67 67 66 66 65 65 64 64 67 67 70 70 71 71 69 69 71 71 68 68 69 69 72 72 73 73 78 78 72 72 78 78 79 79 76 76 77 77 75 75 77 77 74 74 75 75
  </triangles>
  <triangles material="surface-style-3920-masonry---concrete-block" count="20">
    <input semantic="VERTEX" source="#representation-5397-vertices" offset="0"/>
    <input semantic="NORMAL" source="#representation-5397-normals" offset="1"/>
    <p>113 113 115 115 114 114 113 113 112 112 115 115 118 118 119 119 117 117 119 119 116 116 117 117 120 120 122 122 127 127 120 120 127 127 121 121
  </triangles>
  <triangles material="surface-style-3896-misc.-air-layers---air-space" count="12">
    <input semantic="VERTEX" source="#representation-5397-vertices" offset="0"/>
    <input semantic="NORMAL" source="#representation-5397-normals" offset="1"/>
    <p>147 147 149 149 148 148 147 147 146 146 149 149 152 152 153 153 151 151 153 153 150 150 151 151 157 157 154 154 155 155 157 157 155 155 156 156
  </triangles>
  <triangles material="surface-style-3920-masonry---concrete-block" count="20">
    <input semantic="VERTEX" source="#representation-5397-vertices" offset="0"/>
    <input semantic="NORMAL" source="#representation-5397-normals" offset="1"/>
    <p>171 171 173 173 172 172 171 171 170 170 173 173 176 176 177 177 175 175 177 177 174 174 175 175 181 181 178 178 179 179 181 181 179 179 180 180
  </triangles>
  <triangles material="surface-style-3935-metal---stud-layer" count="28">
    <input semantic="VERTEX" source="#representation-5397-vertices" offset="0"/>
    <input semantic="NORMAL" source="#representation-5397-normals" offset="1"/>
    <p>205 205 207 207 206 206 205 205 204 204 207 207 210 210 211 211 209 209 211 211 208 208 209 209 216 216 217 217 214 214 216 216 214 214 215 215
  </triangles>
  <triangles material="surface-style-3944-plasterboard" count="44">
    <input semantic="VERTEX" source="#representation-5397-vertices" offset="0"/>
    <input semantic="NORMAL" source="#representation-5397-normals" offset="1"/>
    <p>256 256 255 255 254 254 253 253 252 252 259 259 254 254 253 253 259 259 254 254 259 259 258 258 254 254 258 258 257 257 254 254 257 257 256 256
  </triangles>
</mesh>
</geometry>
~~~

**Comment by Stinkfist0** [comment]
OK, this is very unfortunate and my hatred towards Unity & Autodesk's FBX SDK continues to grow day by day :D. I'll try to find a workaround.

**Comment by aothms** [comment]
Maybe it would already be sufficient to add multiple `<instance_material>` nodes that bind to the same `target`, e.g post-fix them with a numeric index.

We'd have to check how unity deals with it and check how other importers deal with it and then see if we can make this change unconditionally or make it another command line option.

**Comment by Stinkfist0** [comment]
Yeah, was thinking some kind of number postfix too. I can confirm that this same problem is also present if using OBJ instead.

**Comment by Stinkfist0** [comment]
I started looking into this now as this is a pretty pressing issue for us.

**Comment by aothms** [comment]
Ok, good to hear this feature is being used more and more in production. I can imagine it to improve visualisation results for your use case. Let me know how it works out for you and what points of improvement you encounter.

**Comment by Stinkfist0** [comment]
OK, found a solution/hack to prevent this occurring when importing DAE to Unity https://github.com/Tridify/IfcOpenShell/commit/af4ccc8a1108a3e5495bbd0387b983659e77b1df

**Comment by Stinkfist0** [comment]
Example output
old: https://dl.dropboxusercontent.com/u/16413694/IfcOpenShell/Duplex_A_20110907-old.dae
new: https://dl.dropboxusercontent.com/u/16413694/IfcOpenShell/Duplex_A_20110907-new.dae
That DAE needs to be cleaned up at least so that duplicate materials and effects are not written blindly for all material occurences, but only for cases that really need them.

**Comment by Stinkfist0** [comment]
But now that I got my hack rushed in (my second batch of vacation days starts today), I'm experiencing quite a bit of problems with the model I'd like to use this feature: approx. 90% of elements simply disappear from the output and I've left only with a handful of sliced walls. The model converts fine without layer set slicing. Any pointers where in the code to research for potential problems? Can the slicing fail silently and it simply skips the element in question?

**Comment by aothms** [comment]
If there is some sort of exception in the process, i.e an opencascade `Standard_ConstructionError`, it could well be that the exception is caught at a level where simply the element gets skipped.

The feature has been released into the wild fairly recently, I wouldn't want to claim it is free of bugs, although it worked on all models I tested on. Especially the non-visual information in models, e.g path connections, material layers, can also sometimes contain flaws in the IFC models themselves.

Care to prepare a minimal example where this fails? A wall would need the material layer set association and also the path connections to neighbouring walls influence how the layers are created.

**Comment by Stinkfist0** [comment]
Can I easily use IfcPython to extract a specific element with all of its dependencies into a new IFC file.

**Comment by Stinkfist0** [comment]
@aothms hi, any tips^?

**Comment by Stinkfist0** [comment]
I can also confirm that the neighboring elements do matter: when trying to isolate element that has failed when converting the whole model, the element is fine when converted individually.

**Comment by aothms** [comment]
Apologies, somehow I completely missed your message.

> Can I easily use IfcPython to extract a specific element with all of its dependencies into a new IFC file?

Yes, you can do something like:

~~~
f1, f2 = ifcopenshell.open("existing_file.ifc"), ifcopenshell.open()
my_wall = f1.by_type("IfcWall")[0]
# For wall and forward references
for x in f1.traverse():
  f2.add(x)
# For material association (layerset)
for assoc in my_wall.HasAssociations:
  f2.add(assoc)
# For neighbouring elements
for path in my_wall.ConnectedFrom + my_wall.ConnectedTo:
  f2.add(path)
f2.write("new_file.ifc")
~~~

This will still miss some things probably, I'm guessing at least color information. But probably it's enough to crash on the geometry.

**Comment by Stinkfist0** [comment]
Thanks!

**Comment by Stinkfist0** [comment]
Also the model seems a bit weird (e.g. windows are curtain walls...), so it might not be the best source input. I'll open up a new issue for potential slicing problems, let's continue strictly the sub-mesh merging issue here.

**Comment by aothms** [comment]
Perfect. On a side note, windows as curtain walls is not uncommon these days. Curtain walls are typically used if the element is decomposed into separate struts and glass panes, which is desirable also to be able to associate distinct propertysets to glazing and frame separately. Kind of a (one of the) design flaws if you ask me.

**Comment by Stinkfist0** [comment]
I can now confirm that e.g. Open Asset Import Library suffers from the same problem (tested importing sliced Duplex using Urho3D editor). Therefore I would propose that the solution for this would be the default behavior for OBJ & DAE conversion.

**Comment by Stinkfist0** [comment]
Cleaned up the code https://github.com/Tridify/IfcOpenShell/commit/c584f604598b761efdef433a8385ae70f2c0b85c 

Example output (`--sew-shells --use-element-guids --use-material-names --generate-uvs --enable-layerset-slicing`):
https://dl.dropboxusercontent.com/u/16413694/IfcOpenShell/Duplex-sliced-old.dae
https://dl.dropboxusercontent.com/u/16413694/IfcOpenShell/Duplex-sliced-new.dae

However, the IfcGeomRepresentation changes needs some revising: e.g. some staircases are split now into 32 sub-meshes which is not intentional.

**Comment by Stinkfist0** [comment]
On the other hand, this kind of splitting could be be handy. E.g. now I could pick individual materials for different parts nicely for this cabinet which would have only one material otherwise:
![image](https://cloud.githubusercontent.com/assets/227876/17780404/e64b24b6-6573-11e6-85b6-6332c1ab4fad.png)

**Comment by aothms** [comment]
I can see it's merits. In particular it preserves the association to individual rep items. But on the other hand, the current mode allows to change the appearance of multiple elements in a predictable way, whereas the new mode makes these changes somewhat inconsistent. At least that is my observation playing with you dae files (thanks for the example) in Blender.

### Old: changing plaster material works consistently on all objects

![old](https://cloud.githubusercontent.com/assets/1096535/17831664/404b8b14-66ef-11e6-8a02-d431edfa40e7.png)

### New: changing plaster material is not that predictable anymore. Changing plasterboard material instances with all possible postfixes is not that easy in Blender.

![new](https://cloud.githubusercontent.com/assets/1096535/17831666/4321cf60-66ef-11e6-8fcf-8a86ffb0bdd7.png)

My inclination would be to make this an optional command line parameter that is off by default.

**Comment by Stinkfist0** [comment]
Yeah, that might make the most sense for the usual cases. I also noticed that the layer set slicing has nothing to do with the issue and I can actually remember wondering about this issue with one model maybe a 1.5 years ago when the slicing was not even yet available.

**Comment by aothms** [comment]
Ok, so how to go about this? Perhaps the most predictable alternative to the current procedure is to make material labels globally unique. That is to say, postfix them with both their index in the material slots for the current element as well as with a unique id for the element. This way every label is assigned only once. I can image this might be useful for other purposes as well, e.g programmatically color coding collada output based on some external data sources. Thoughts?

**Comment by Stinkfist0** [comment]
Hmm so you would suggest something similar that my first hack somewhat did, i.e. every occurrence of a material would generate a unique material instance in the output (DAE or OBJ)?

**Comment by aothms** [comment]
Didnt realize this was similar to your first approach. But indeed, think that would be most beneficial to have as an alternative to what exists now.

**Comment by Stinkfist0** [comment]
Yeah, halfway solution (make some materials unique and some not) might be confusing. So this feature might be something like "enforce unique materials".

**Comment by Stinkfist0** [comment]
I will look into finishing this next.

**Comment by Stinkfist0** [comment]
@aothms Would you have a preference if the `_<n>` postfix is applied always or only when there are more than one references to a material in a model?

**Comment by aothms** [comment]
My mind is a bit fuzzy on this one. So this is with an `--enforce-unique-materials` option enabled? Perhaps always postfix then to make things the most predictable?

**Comment by aothms** [comment]
Then, perfect, I would always apply the postfix in that mode.

**Comment by Stinkfist0** [comment]
After giving this more though, I probably will end up writing a simple XML/DAE processor tool/script which fixes this as bloating already bloated DAE doesn't feel that nice.

**Comment by aothms** [comment]
Your call. If we would have used a generic XML lib instead of OpenCOLLADA, perhaps it would be possible to do sth with XSLT or other magic, to keep concerns separated.

**Comment by Moult** [comment]
Is this resolved? Materials and styles now have unique ID/names, yes?

**Comment by aothms** [comment]
This was very specific to Unity. It was decided not to address this in IfcConvert.
